lxd/events.go

```diff
@@ -162,6 +180,58 @@ func eventsSocket(s *state.State, r *http.Request, w http.ResponseWriter) error
 		}
 	}
 
+	filter := func(log logger.Logger, event api.Event) bool {
+		l = log.AddContext(logger.Ctx{"type": event.Type, "location": event.Location, "project": event.Project})
+
+		// Privileged events require `can_view_events` on `server.
+		if slices.Contains(privilegedEventTypes, event.Type) {
+			return canViewServerEvents
+		}
+
+		switch event.Type {
+		case api.EventTypeLifecycle:
+			// Lifecycle events that are not project specific require `can_view_events` on `server`.
+			if event.Project == "" {
+				return canViewServerEvents
+			}
+
+			// Otherwise check if the caller has `can_view_lifecycle_events` on the project.
+			if canViewProjectLifecycleEvents(entity.ProjectURL(event.Project)) {
+				return true
+			}
+
+		case api.EventTypeOperation:
+			// Operations that are not project specific require `can_view_operations` on `server`.
+			if event.Project == "" {
+				return canViewServerOperations
+			}
+
+			// Otherwise check if the caller has `can_view_operations` on the project.
+			if canViewProjectOperations(entity.ProjectURL(event.Project)) {
+				return true
+			}
+
+		default:
+			// We don't expect any other event types at this point
+			l.Warn("Received unexpected event type")
+			return false
+		}
+
+		// At this point the caller does not have permission to view the event via group membership or project.
+		// Check the event or operation requestor to see if they are the identity that triggered the event.
+
+		// Unmarshal the requestor from the event metadata.
+		var m requestorMetadata
+		err := json.Unmarshal(event.Metadata, &m)
+		if err != nil {
+			l.Error("Failed to unmarshal event metadata during requestor filtering")
+			return false
+		}
+
+		// Allow the event if the same requestor is connected. Otherwise, filter it out.
+		return m.Requestor != nil && m.Requestor.Username == requestor.CallerUsername() && m.Requestor.Protocol == requestor.CallerProtocol()
```

Comment:
@markylaing just a nit, not a blocker, but im not a fan of these special case return statements at the end of a function, prefer to have each condition a separate `if` block with a simple default return value at the bottom. This makes it easier to refactor the function in the future in my experience.